### PR TITLE
source aha.io: fix for optional streams

### PR DESCRIPTION
**Pull request by upgreydd**
## What
`features` and `goals` are optional - while creating an integration connection check based on `features`, but it should base on `products`.
## How
Changed order of streams to check on `products` instead of `features`

## 🚨 User Impact 🚨
Fixing aha.io synchronization on lower plans where is no `Features` and `Goals` enabled by aha.io

### CODE REVIEW

File: airbyte-integrations/connectors/source-aha/acceptance-test-docker.sh [hunk 1]
```diff
@@ -1,7 +1,7 @@
 #!/usr/bin/env sh
 
 # Build latest connector image
-docker build . -t $(cat acceptance-test-config.yml | grep "connector_image" | head -n 1 | cut -d: -f2-)
```
Comment (Upgreydd) [inline airbyte-integrations/connectors/source-aha/acceptance-test-docker.sh:4]: ShellCheck fix

File: airbyte-integrations/connectors/source-aha/acceptance-test-docker.sh [hunk 2]
```diff
@@ -10,7 +10,7 @@ docker pull airbyte/connector-acceptance-test:latest
 docker run --rm -it \
     -v /var/run/docker.sock:/var/run/docker.sock \
     -v /tmp:/tmp \
-    -v $(pwd):/test_input \
```
Comment (Upgreydd) [inline airbyte-integrations/connectors/source-aha/acceptance-test-docker.sh:13]: ShellCheck fix

File: airbyte-integrations/connectors/source-aha/integration_tests/configured_catalog.json [hunk 3]
```diff
@@ -1,14 +1,5 @@
...
   "streams": [
-    {
-      "stream": {
-        "name": "features",
-        "json_schema": {},
-        "supported_sync_modes": ["full_refresh"]
-      },
-      "sync_mode": "full_refresh",
-      "destination_sync_mode": "overwrite"
-    },
```
Comment (Upgreydd) [inline airbyte-integrations/connectors/source-aha/integration_tests/configured_catalog.json:11]: removal of optional stream

File: airbyte-integrations/connectors/source-aha/integration_tests/configured_catalog.json [hunk 4]
```diff
@@ -35,15 +26,6 @@
...
       "destination_sync_mode": "overwrite"
-    },
-    {
-      "stream": {
-        "name": "goals",
-        "json_schema": {},
-        "supported_sync_modes": ["full_refresh"]
-      },
-      "sync_mode": "full_refresh",
-      "destination_sync_mode": "overwrite"
```
Comment (Upgreydd) [inline airbyte-integrations/connectors/source-aha/integration_tests/configured_catalog.json:46]: removal of optional stream

File: airbyte-integrations/connectors/source-aha/source_aha/manifest.yaml [hunk 5]
```diff
@@ -25,6 +25,10 @@ definitions:
     authenticator:
       type: "BearerAuthenticator"
       api_token: "{{ config['api_key'] }}"
+    error_handler:
+      response_filters:
+        - error_message_contains: "Your role is"
+          action: IGNORE
```
Comment (Upgreydd) [inline airbyte-integrations/connectors/source-aha/source_aha/manifest.yaml:31]: handling optional streams

File: airbyte-integrations/connectors/source-aha/source_aha/source.py [hunk 6]
```diff
@@ -13,6 +21,49 @@
 
 
 # Declarative Source
-class SourceAha(YamlDeclarativeSource):
+class Aha(YamlDeclarativeSource):
     def __init__(self):
         super().__init__(**{"path_to_yaml": "manifest.yaml"})
+
+
+def check_existence(config: Mapping[str, Any], url_param: str) -> Union[tuple[bool, None], tuple[bool, str]]:
```
Comment (Upgreydd) [inline airbyte-integrations/connectors/source-aha/source_aha/source.py:29]: for checking existance of stream - as I said in PR - it's returning `403` with listed roles

File: airbyte-integrations/connectors/source-aha/source_aha/source.py [hunk 7]
```diff
@@ -13,6 +21,49 @@
...
+    response = requests.head(urljoin(config['url'], f"/api/v1/{url_param}"), headers={"Authorization": f"Bearer {config['api_key']}"})
+    if response.status_code == 200:
+        return True, None
+    return False, response.text
+
+
+class AhaEmptyBase(Stream):
```
Comment (Upgreydd) [inline airbyte-integrations/connectors/source-aha/source_aha/source.py:36]: Class for introducing empty records

File: airbyte-integrations/connectors/source-aha/source_aha/source.py [hunk 8]
```diff
@@ -13,6 +21,49 @@
...
+    primary_key = None
+
+    def read_records(self, sync_mode: SyncMode, cursor_field: List[str] = None, stream_slice: Mapping[str, Any] = None,
+                     stream_state: Mapping[str, Any] = None) -> Iterable[StreamData]:
+        return []
+
+
+class Goals(AhaEmptyBase):
```
Comment (Upgreydd) [inline airbyte-integrations/connectors/source-aha/source_aha/source.py:44]: Class to override goals when they doesn't exists

File: airbyte-integrations/connectors/source-aha/source_aha/source.py [hunk 9]
```diff
@@ -13,6 +21,49 @@
...
+    pass
+
+
+class Features(AhaEmptyBase):
```
Comment (Upgreydd) [inline airbyte-integrations/connectors/source-aha/source_aha/source.py:48]: Class to override features when they don't exists

File: airbyte-integrations/connectors/source-aha/source_aha/source.py [hunk 10]
```diff
@@ -13,6 +21,49 @@
...
+    pass
+
+
+class SourceAha(AbstractSource):
+    def check_connection(self, logger: logging.Logger, config: Mapping[str, Any]) -> Tuple[bool, Optional[Any]]:
+        # check the connection via `product` because it's required element on the AHA.IO side. `Features` endpoint can return 403
+        return check_existence(config, 'products')
+
+    def streams(self, config: Mapping[str, Any]) -> List[Stream]:
```
Comment (Upgreydd) [inline airbyte-integrations/connectors/source-aha/source_aha/source.py:57]: fixing streams if optional streams doesn't exist

File: airbyte-integrations/connectors/source-aha/source_aha/source.py [hunk 11]
```diff
@@ -13,6 +21,49 @@
...
+
+class Goals(AhaEmptyBase):
+    pass
+
+
+class Features(AhaEmptyBase):
+    pass
+
+
+class SourceAha(AbstractSource):
```
Comment (Upgreydd) [inline airbyte-integrations/connectors/source-aha/source_aha/source.py:52]: checking connection state by required element, not optional

File: airbyte-integrations/connectors/source-aha/acceptance-test-docker.sh [hunk 12]
```diff
@@ -1,7 +1,7 @@
 #!/usr/bin/env sh
 
 # Build latest connector image
-docker build . -t $(cat acceptance-test-config.yml | grep "connector_image" | head -n 1 | cut -d: -f2-)
+docker build . -t "$(< acceptance-test-config.yml grep "connector_image" | head -n 1 | cut -d: -f2- | xargs)"
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-aha/acceptance-test-docker.sh:4]: Why are you doing those changes and changing the file permission from 100644 -> 100755?

File: airbyte-integrations/connectors/source-aha/integration_tests/configured_catalog.json [hunk 13]
```diff
@@ -1,14 +1,5 @@
 {
   "streams": [
-    {
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-aha/integration_tests/configured_catalog.json:3]: The configured catalog is used to run tests. If your account doesn't have permissions just remove it, but don't remove it Airbyte will test them.

File: airbyte-integrations/connectors/source-aha/source_aha/source.py [hunk 14]
```diff
@@ -13,6 +21,49 @@
...
+                     stream_state: Mapping[str, Any] = None) -> Iterable[StreamData]:
+        return []
+
+
+class Goals(AhaEmptyBase):
+    pass
+
+
+class Features(AhaEmptyBase):
+    pass
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-aha/source_aha/source.py:49]: I don't believe the change you're doing it is correct.

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
Sorry @Upgreydd the changes you're going doesn't look correct. Check comments. If your account doesn't have permissions to use determined stream don't select them to sync.
I strongly recommend to open first an issue, describe the problem post the error log and after submit the fix.

File: airbyte-integrations/connectors/source-aha/acceptance-test-docker.sh (hunk 12 above)
Comment (Upgreydd) [reply airbyte-integrations/connectors/source-aha/acceptance-test-docker.sh:4]: OK I'll revert that part - it's more capable with ShellCheck now, but as you wish

File: airbyte-integrations/connectors/source-aha/integration_tests/configured_catalog.json (hunk 13 above)
Comment (Upgreydd) [reply airbyte-integrations/connectors/source-aha/integration_tests/configured_catalog.json:3]: OK understood

File: airbyte-integrations/connectors/source-aha/source_aha/source.py (hunk 14 above)
Comment (Upgreydd) [reply airbyte-integrations/connectors/source-aha/source_aha/source.py:49]: > I don't believe the change you're doing it is correct.

Can you evaluate that, please?

Comment (marcosmarxm) [reply airbyte-integrations/connectors/source-aha/source_aha/source.py]: The aha.io API docs https://www.aha.io/api/resources/features/list_features

Are you trying to develop something with this connector? If so, only remove from the configured catalog and it should work fine for your case.

File: airbyte-integrations/connectors/source-aha/source_aha/source.py [hunk 15]
```diff
@@ -16,3 +19,9 @@
 class SourceAha(YamlDeclarativeSource):
     def __init__(self):
         super().__init__(**{"path_to_yaml": "manifest.yaml"})
+
+    def check_connection(self, logger, config) -> Tuple[bool, any]:
+        response = requests.head(urljoin(config['url'], f"/api/v1/products"), headers={"Authorization": f"Bearer {config['api_key']}"})
+        if response.status_code == 200:
+            return True, None
+        return False, response.text
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-aha/source_aha/source.py:27]: Aha Source already uses the `check_connection` from the low-code CDK.

https://github.com/airbytehq/airbyte/blob/a6990bc1958c28fcc4718676f8fc5f9710cdb108/airbyte-integrations/connectors/source-aha/source_aha/manifest.yaml#L103-L109

**Review by marcosmarxm — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-aha/source_aha/manifest.yaml [hunk 16]
```diff
@@ -102,8 +102,8 @@ streams:
 
 check:
   stream_names:
-    - "features"
     - "products"
+    - "features"
     - "ideas"
     - "users"
     - "goals"
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-aha/source_aha/manifest.yaml:109]: @Upgreydd you want to remove the goals and features from check connection?

Comment (Upgreydd) [reply airbyte-integrations/connectors/source-aha/source_aha/manifest.yaml:109]: @marcosmarxm looks like connection check base only on first element, when I changed order from optional `features` first to `products` - then it pass.

**Comment by Upgreydd** [comment]
> @Upgreydd please create an issue describing what problem are you having today.

https://github.com/airbytehq/airbyte/issues/22645

**Comment by marcosmarxm** [comment]
@archangelic to get this done:
- [ ] update the manifest.yaml to latest version (I think is 0.29.0)
- [ ] change the old `*ref` to new one `#ref/..`
- [ ] make sure schemas are updated
- [ ] integration tests should be passing now
- [ ] publish

**Comment by marcosmarxm** [comment]
This was fixed in https://github.com/airbytehq/airbyte/pull/26666

**Pull request closed without merging**